### DISCUSSION TITLE
Partial const fields

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5893] [Language Ideas]
This idea hit me when thinking about the common pattern of declaring const strings:
```csharp
public const string SomeProperty = nameof(SomeProperty);
```

This process could be automated via source generators that could make use of partial const field declarations. This should work exactly like partial methods; one signature declaration and one implementation declaration.

W.r.t. the above example, this would act like so:

The user writes the following code (assume `NamedConstStrings` triggers the source generator)
```csharp
[NamedConstStrings]
public partial static class PropertyNames
{
    public partial const string
        SomeProperty,
        SomeOtherProperty,
        Value,
        Date;
}
```

And the generator generates the following code:
```csharp
public partial static class PropertyNames
{
    public partial const string SomeProperty = nameof(SomeProperty);
    public partial const string SomeOtherProperty = nameof(SomeOtherProperty);
    public partial const string Value = nameof(Value);
    public partial const string Date = nameof(Date);
}
```

----

Another example is simulating `constexpr` in C++, by passing down an expression in a string form and generating the constant result of that expression, assigning it to the field. For instance,
```csharp
[ConstExpr("Math.Max(4, 5)")]
public partial const int Max45;
```

in which case the generator would parse the expression, execute it and store the produced result to the initialization part of the declaration:
```csharp
public partial const int Max45 = 5;
```

This specific example might be too far-fetched for certain scenarios, though it could eliminate some of the existing `const` limiations, until a better solution is found in future versions of the language.

**Comment by huoyaoyuan** [comment]
Partial method is used for code generation because the generator needs full method signature. It's hard to encode method signature in other way.
In contrast, encoding a string constant in an attribute is easy.

**Reply by Rekkonnect** [reply]
From what I understand, what you're trying to say is that partial const fields aren't necessary as you can declare those string fields in raw strings in attributes, like so:
```csharp
[ConstStrings(
"Property",
"OtherProperty",
...)]
```

Not only does it look ugly, it also requires heavy unfortunate usage of the `"` character. Plus, this only covers this one case. There's more cases for const field generators, which I'll include in the original thread.
